src/components/search/SearchTypesDirectives.js

```diff
@@ -0,0 +1,395 @@
+        canceler.resolve();
+        canceler = undefined;
+      }
+    };
+
+    var triggerSearch = function() {
+      if (!$scope.doSearch()) {
+        $scope.options.announceResults(type, 0);
+        return;
+      }
+
+      canceler = $q.defer();
+
+      var url = urlUtils.append($scope.options.baseUrl, 'type=' + type);
+      url = $scope.typeSpecificUrl(url);
+
+      $http.get(url, {
+        cache: true,
+        timeout: canceler.promise
+      }).success(function(data) {
+        $scope.results = data.results;
+        $scope.options.announceResults(type, data.results.length);
+      }).error(function(data, statuscode) {
+        // If request is canceled, statuscode is 0 and we don't announce it
+        if (statuscode !== 0) {
+          $scope.options.announceResults(type, 0);
+        }
+      });
+    };
+
+    $scope.type = type;
+
+    $scope.doSearch = function() {
+      return true;
+    };
+
+    $scope.typeSpecificUrl = function(url) {
+      return url;
+    };
+
+    $scope.preview = function(res) {
+      if (browserSniffer.mobile) {
+        return;
+      }
+      addOverlay(markerOverlay, $scope.map, res);
+    };
+
+    $scope.removePreview = function(evt, res) {
+      if (browserSniffer.mobile ||
+          // HACK for #1737: prevent unexpected call of mouseout evt.
+          // TODO: maybe this hack is not needed anymore with
+          // angular only search
+          (evt.target != evt.relatedTarget &&
+          evt.relatedTarget instanceof HTMLCanvasElement)) {
+        return;
+      }
+      removeOverlay(markerOverlay, $scope.map);
+    };
+
+    $scope.prepareLabel = function(attrs) {
+      var h = searchLabels.highlight(attrs.label, $scope.options.query);
+      return $sce.trustAsHtml(h);
+    };
+
+    $scope.$watch('options.query', function(newval) {
+      //cancel old requests
+      cancel();
+      if (newval != '') {
+        triggerSearch();
+      } else {
+        unregisterMove();
+      }
+    });
+
+  };
+
+  var module = angular.module('ga_search_type_directives', [
+    'ga_browsersniffer_service',
+    'ga_debounce_service',
+    'ga_layer_metadata_popup_service',
+    'ga_map_service',
+    'ga_marker_overlay_service',
+    'ga_search_service',
+    'ga_urlutils_service',
+    'pascalprecht.translate'
+  ]);
+
+  module.directive('gaSearchLocations',
+      function($http, $q, $sce, $translate, gaUrlUtils, gaBrowserSniffer,
+               gaMarkerOverlay, gaSearchLabels, gaMapUtils, gaDebounce) {
+        return {
+          restrict: 'A',
+          replace: true,
+          templateUrl: 'components/search/partials/searchtypes.html',
+          scope: {
+            options: '=gaSearchLocationsOptions',
+            map: '=gaSearchLocationsMap'
+          },
+          controller: function($scope) {
+            setupController($scope, 'locations', $http, $q, $sce, gaUrlUtils,
+                            gaSearchLabels, gaBrowserSniffer, gaMarkerOverlay);
```

Comment:
Had it early on and ran into troubles, but changed architecture considerably since, so it might be good now. I'l give it a try, it's the ugliest part of the code anyway.